packages/logger/src/logger.ts

```diff
@@ -137,23 +147,82 @@ export class Logger {
     this._logHandler = val;
   }
 
+  /**
+   * The optional, additional, user-defined log handler for the Logger instance.
+   */
+  private _userLogHandler: LogHandler | null = null;
+  get userLogHandler(): LogHandler | null {
+    return this._userLogHandler;
+  }
+  set userLogHandler(val: LogHandler | null) {
+    this._userLogHandler = val;
+  }
+
   /**
    * The functions below are all based on the `console` interface
    */
 
   debug(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.DEBUG, ...args);
     this._logHandler(this, LogLevel.DEBUG, ...args);
   }
   log(...args: unknown[]): void {
+    this._userLogHandler &&
+      this._userLogHandler(this, LogLevel.VERBOSE, ...args);
     this._logHandler(this, LogLevel.VERBOSE, ...args);
   }
   info(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.INFO, ...args);
     this._logHandler(this, LogLevel.INFO, ...args);
   }
   warn(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.WARN, ...args);
     this._logHandler(this, LogLevel.WARN, ...args);
   }
   error(...args: unknown[]): void {
+    this._userLogHandler && this._userLogHandler(this, LogLevel.ERROR, ...args);
     this._logHandler(this, LogLevel.ERROR, ...args);
   }
 }
+
+export function setLogLevel(level: LogLevelString | LogLevel): void {
+  const newLevel = typeof level === 'string' ? levelStringToEnum[level] : level;
+  instances.forEach(inst => {
+    inst.logLevel = newLevel;
+  });
+}
+
+export function setUserLogHandler(
+  logCallback: LogCallback | null,
+  options: LogOptions
+): void {
+  if (typeof logCallback !== 'function') {
+    console.warn('First argument to `onLog` must be a function.');
+    return;
+  }
+  for (const instance of instances) {
+    let threshhold = instance.logLevel;
+    if (options && options.level) {
+      threshhold = levelStringToEnum[options.level];
+    }
+    if (logCallback === null) {
+      instance.userLogHandler = null;
+    } else {
+      instance.userLogHandler = (
+        instance: Logger,
+        level: LogLevel,
+        ...args: unknown[]
+      ) => {
+        const message = args.map(arg => (arg as object).toString()).join(' ');
```

Comment:
Looked at how we use logger internally and thought about it again and , I think we should just assign `args[0]` to `message` and pass the rest args as is.

`toString()` produces `object [Object]` for objects, which is not very helpful. WDYT?

Reply:
We would have to enforce the convention of putting the entire message string in the first arg which I guess is possible, there are only a few existing cases where it wouldn't work, such as https://github.com/firebase/firebase-js-sdk/blob/542240d093ad54f3c8e5739dcb8449e8b9cab131/packages/firestore/src/local/indexeddb_persistence.ts#L1018.  In that case the message should be the first 2 args.  It might be easier to change the few existing lines in Firestore that don't fit, and enforce the convention going forward, than to handle all cases.

If we decided to do a robust handling of all args, I was thinking we could check arg type and do `toString()` on primitives, and `JSON.stringify()` on arrays and objects.